ts/nix.ts

```diff
@@ -132,37 +151,148 @@ export function nixStrLit(
-            );
+  if (typeof s === "string") {
+    return { __ast: { type: "strLit", str: interpolatedStringFromString(s) } };
+  }
+  return {
+    __ast: {
+      type: "strLit",
+      str: interpolatedStringFromTemplate(
+        s,
+        interpolations.map((interpolation): NixAst => {
+          if (typeof interpolation === "string") {
+            return {
+              type: "strLit",
+              str: interpolatedStringFromString(interpolation),
+            };
           }
-          return (
-            acc +
-            part +
-            "${" +
-            interpolation.nixExpression.rawNixExpressionString +
-            "}"
+          if ("__ast" in interpolation) {
+            return interpolation.__ast;
+          }
+          return interpolation.nixExpression.__ast;
+        }),
+      ),
+    },
+  };
+}
+
+/**
+ * Converts a `NixExpression` to a (hopefully concise) human readable string
+ * with incidental dependencies snipped out with "[...]"
+ */
+export function toHumanReadable(nixExpr: NixExpression): string {
+  const astToHumanReadable = (node: NixAst): string => {
+    if (node.type !== "strLit") {
+      return "[...]";
+    }
+    return node.str.rest.reduce(
+      (acc, [node, str]) => acc + astToHumanReadable(node) + str,
+      node.str.initial,
+    );
+  };
+  return astToHumanReadable(nixExpr.__ast);
+}
+
+/**
+ * Converts a `NixExpression` to a string representing that Nix expression.
+ *
+ * It is advised to defer calling this unless actually serializing to disk.
+ * This way Nix expressions remain as the type `NixExpression`.
+ */
+export function toNixString(nixExpr: NixExpression): string {
+  const astToNixString = (node: NixAst): string => {
+    switch (node.type) {
+      case "raw":
+        return node.raw.rest
+          .reduce(
+            (acc, [astNode, str]) => acc + astToNixString(astNode) + str,
+            node.raw.initial,
+          )
+          .trim();
+      case "list":
+        return (
+          "[" + node.elements.map((e) => astToNixString(e)).join(" ") + "]"
+        );
+      case "attrSet":
+        return (
+          "{ " +
+          Object.entries(node.elements)
+            .map(
+              ([k, v]) =>
+                `${astToNixString(nixStrLit(k).__ast)} = ${astToNixString(v)};`,
+            )
+            .join(" ") +
+          " }"
+        );
+      case "strLit": {
+        const escape = (str: string) =>
+          ["\\", '"', "$"].reduce(
+            (str, char) => str.replaceAll(char, `\\${char}`),
+            str,
           );
-        case "undefined":
-          return acc + part;
+        return (
+          '"' +
+          node.str.rest.reduce(
+            (acc, [astNode, str]) =>
+              acc + "${" + astToNixString(astNode) + "}" + escape(str),
+            escape(node.str.initial),
+          ) +
+          '"'
+        );
       }
-    }, "") +
-    '"';
-  return { rawNixExpressionString };
+    }
```

Comment:
Can we use `utils.checkExhaustiveness` here to make sure we cover all ast node types?

Reply:
Yeah, although the return type of this function does enforce this, `checkExhaustiveness` probably gives a better type error location though

Reply:
Oh, I see. Never mind then!